REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [4]:
import pandas as pd

In [2]:
!conda install pytables --yes

Solving environment: done

# All requested packages already installed.



In [5]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [4]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [5]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [24]:
bus.describe(include='all')

print("\033[1m"+'There are 6315 records'+"\033[1m")

There are 6315 records


#### Question 2b: How many unique business IDs are there?  

In [25]:
bus['business_id'].unique()
print("\033[1m"+'There are 6315 unique businesses'+"\033[1m")

There are 6315 unique businesses


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [26]:
print(bus['name'].value_counts())

print("\033[1m"+'#1 Starbucks:72, #2 Peets:24, #3 Macdos:12, #4 SF Soup:11, #5 Walgreens:11'+"\033[1m")

STARBUCKS COFFEE                      72
PEET'S COFFEE & TEA                   24
MCDONALDS                             12
SAN FRANCISCO SOUP COMPANY            11
WALGREENS                             11
                                      ..
LONGFELLOW ELEMENTARY SCHOOL           1
SLIDER DINER                           1
ALICE'S RESTAURANT                     1
HONEY B, INC.                          1
RICHMOND SENIOR CENTER - 3RD FLOOR     1
Name: name, Length: 5692, dtype: int64
#1 Starbucks:72, #2 Peets:24, #3 Macdos:12, #4 SF Soup:11, #5 Walgreens:11


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [10]:
zipcode = bus['postal_code']
randzip = zipcode[10]
print(randzip)
type(randzip)

94133


str

In [11]:
print('It is a string')

It is a string


#### Question 3b

What are the unique values of postal_code?

In [12]:
bus.postal_code.unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [2]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [6]:
vb = bus[bus['postal_code'].isin(validZip)]
print(vb.describe(include='all'))
print("\033[1m"+'There are 5438 businesses'+"\033[1m")

         business_id              name                 address           city  \
count    5999.000000              5999                    5999           5999   
unique           NaN              5438                    5345              1   
top              NaN  STARBUCKS COFFEE  24 WILLIE MAYS PLAZA    San Francisco   
freq             NaN                71                     122           5999   
mean    47485.367061               NaN                     NaN            NaN   
std     33977.547969               NaN                     NaN            NaN   
min        10.000000               NaN                     NaN            NaN   
25%      5798.000000               NaN                     NaN            NaN   
50%     63943.000000               NaN                     NaN            NaN   
75%     78780.000000               NaN                     NaN            NaN   
max     90239.000000               NaN                     NaN            NaN   

       state postal_code   

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [16]:
vb[pd.isnull(vb['longitude'])]
print('2483 do not have a valid longitude')

2483 do not have a valid longitude


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [19]:
vblon = vb[vb['longitude']<0]
vblon.rename(columns={'longitude':'not_null_lon'},inplace=True)
vbnul = vb[vb['longitude'].isnull()]
vbnul['longitude']=vbnul['longitude'].fillna(0)
vbnul.rename(columns={'longitude':'null_lon'},inplace=True)
vblong = vblon[['postal_code','not_null_lon']].groupby(['postal_code']).count()
vbnulg = vbnul[['postal_code','null_lon']].groupby(['postal_code']).count()

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4441: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(
<ipython-input-19-968127fb6f9a>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vbnul['longitude']=vbnul['longitude'].fillna(0)


In [20]:
vbg=pd.merge(vblong,vbnulg,on=['postal_code'])
vbg

not_null_lon  null_lon
postal_code                        
94102                 262       196
94103                 306       252
94104                  74        59
94105                 126       100
94107                 204       247
94108                 142        82
94109                 237       146
94110                 334       251
94111                 167       115
94112                 131        59
94114                 122       108
94115                 162        76
94116                  58        35
94117                 130        67
94118                 143       100
94121                 106        55
94122                 147       123
94123                 116        53
94124                  76        88
94127                  43        28
94131                  34        14
94132                  69        65
94133                 282       135
94134                  45        29

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [21]:
vbg['ratio']=vbg['null_lon']/(vbg['not_null_lon']+vbg['null_lon'])
vbs=vbg.sort_values('ratio',ascending=False)
vbs
print('Top 5 Postal Codes with null longitudes: 1) 94107, 2) 94124, 3) 94132, 4) 94114, 5) 94122')

Top 5 Postal Codes with null longitudes: 1) 94107, 2) 94124, 3) 94132, 4) 94114, 5) 94122


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [22]:
ins['business_id'].describe()
print('There are 5730 unique businesses')

There are 5730 unique businesses


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [23]:
ins['type'].describe()
ins['type'].unique()
print('There are 2 values that Type takes on: Routine and Complaint')
print(ins[ins['type']=='routine'].count())
print(ins[ins['type']=='complaint'].count())
print("\033[1m"+'There are 15429 Routine inspections and 1 Complaint inspection'+"\033[1m")

There are 2 values that Type takes on: Routine and Complaint
business_id    15429
score          15429
date           15429
type           15429
dtype: int64
business_id    1
score          1
date           1
type           1
dtype: int64
There are 15429 Routine inspections and 1 Complaint inspection


In [27]:
ins1 = ins[['business_id','type']].groupby('business_id').count()
ins1.describe()
ins2 = ins1[ins1['type']>=10]
ins2

Empty DataFrame
Columns: [type]
Index: []

In [28]:
print("\033[1m"+'There are no businesses with 10 or more inspections of the same type.  Here is the frame with 7 or more inspections of the same type to confirm that it is not a code issue'+"\033[1m")
ins2a = ins1[ins1['type']>=7]
ins2a

There are no businesses with 10 or more inspections of the same type.  Here is the frame with 7 or more inspections of the same type to confirm that it is not a code issue


type
business_id      
1516            8
5450            7
5455            8
5461            8
5470            7
5471            8
5478            7
5479            8
5486            7
5487            8
5495            7
5497            8
5499            8
5500            7
5626            8
5628            8
5633            8
5634            7
5641            7
5643            7
5644            7
5649            8
5791            7
5794            7
5831            7
5849            7
5871            7
5902            7
10377           7
69048           7
74336           7
74760           7
76230           7
76233           7
76355           8
78372           8
79905           7
79906           7

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [29]:
type(ins['date'])
insy = ins
insy['date']=ins['date'].astype(str)
insy['year']=insy['date'].str[0:4]
insy.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [30]:
insg=insy[['type','year']].groupby(['year']).count()
insg

type
year      
2013    38
2014  5629
2015  3923
2016  5840

In [31]:
ins3 = insg[insg['type']>50]
ins3

type
year      
2014  5629
2015  3923
2016  5840

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [32]:
ins2016=insy[insy['year']=='2016']
ins16g=ins2016[['score','business_id']].groupby('business_id').mean()
ins16g.head()

score
business_id       
10            82.0
19            94.0
24            97.0
31            84.0
45            81.0

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [33]:
busins = pd.merge(bus,ins16g,on='business_id',how='left')
busins.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  score  
0 -122.403816  +14154217044   82.0  
1 -122.421547  +14157763262   94.0  
2 -122.403135  +14156779494   97.0  
3 -122.419004           NaN   84.0  
4 -122.413641  +14156415051   81.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [34]:
busins16 = busins[['business_id','name','address','score']]
busins16=busins16.sort_values('score',ascending=True)
busins16.head(20)

business_id                         name              address  score
3969        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
3952        71232                   UNCLE CAFE       65 WAVERLY PL    55.0
5118        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
6034        87652                    POKI TIME     2101 LOMBARD ST    55.0
2373        34585             HING WANG BAKERY        339 JUDAH ST    56.0
4587        78328                   GOLDEN WOK   295 B ORIZABA AVE    56.0
3399        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
3978        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
4938        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
5439        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
5702        85587         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3424        67564                        ITHAI         720 POST ST    59.0
2842        61161                  JADE GARDEN      3608 BALBOA ST    59.0
4072        72176                 WOLFES LUNCH        1220 16TH ST    60.0
5009        81106          VALLARTA'S TACO BAR     2022 MISSION ST    60.0
662          2622         WING HING RESTAURANT  2550 SAN BRUNO AVE    61.0
572          2263            TAIWAN RESTAURANT      445 CLEMENT ST    61.0
3750        69512               ELEPHANT SUSHI        1916 HYDE ST    61.0
5735        85835             NEW WORLD MARKET     5641 GEARY BLVD    61.0
4887        80273           TIKKA MASALA ELLIS        425 ELLIS ST    62.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!